### ISSUE TITLE
"destructured" option for camelCase rule

### ISSUE THREAD

**Issue by willrstern** [eslint/eslint#3185]
when destructuring an underscore_spaced request:
`const { category_id: category } = query;`
camelCase errors that `category_id` is not in camel case

It would be nice to have a "destructuring" option, or just ignore when destructuring non-camel-case props to camelCase.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by btmills** [comment]
Since the pattern key is "external" code*, my first reaction is that we should be ignoring that by default, no option necessary. Thoughts?

--

*If it comes from a response, then it's truly external. If it comes from an object created elsewhere, camelcase will catch the issue then. Either way, the local variable is what matters.

**Comment by willrstern** [comment]
Sorry, I meant to say "request" not "response".
In this use case, I'm destructuring a POST request query - so as to only have camelCased locals.

I can do it without destructuring: 
`const category = query.category_id;`
just not with:
`const { category_id: category } = query;`

**Comment by btmills** [comment]
Thanks for the extra context.

```js
/* eslint camelcase: 2 */
var category = query.category_id;
```

The above is valid per the demo, so I'm even more convinced now that its destructuring equivalent should also be considered valid.

**Comment by nzakas** [comment]
I think this should be covered by the `properties` option, shouldn't it?
http://eslint.org/docs/rules/camelcase#options

**Comment by willrstern** [comment]
The `properties` option would probably cover it, but it's really covering up the fact that parsing isn't distinguishing between destructuring-renaming from object-property-naming. 

Adding `properties` would also stop camelCasing checks on all other properties throughout the project.

here's a more simple use-case:
```
import somemodule from "somemodule";

const { some_value: someValue } = somemodule; //errors though no underscore_spaced var was set
const someValue = somemodule.some_value; //works
```
.eslintrc
```
{
  "ecmaFeatures": {
    "destructuring": true,
    "modules": true,
  },

  "env": {
    "es6": true
  },
}
```

**Comment by nzakas** [comment]
Hmm, yeah, we probably need to distinguish between MemberExpression and Property nodes rather than treating them the same.

**Comment by FunkMonkey** [comment]
This still seems to be an issue when destructuring imports, f. ex.:

```js
import { foo_bar } from 'foo';
```
It is especially problematic as the user has no control over the naming in external libraries.

**Comment by alberto** [comment]
@FunkMonkey can you open a new issue following our guidelines, please?

**Comment by mathieumg** [comment]
I don't see the new issue linked here, but @FunkMonkey couldn't you do:

```js
import { foo_bar as fooBar } from 'foo';
```
